File: center/receiver/tbf.py

class TokenBucketFilter:
    """ A Token Bucket Filter """
    def __init__(self, rate, burst, timefunc):
        self._rate = rate
        self._burst = burst
        self._cap = burst
        self._timefunc = timefunc
        self._now = self._timefunc()

    def _replenish(self):
        now = self._timefunc()
        self._cap += (now - self._now) * self._rate
        self._cap = min(self._cap, self._burst)
        self._now = now

    def feed(self, amount):
        self._replenish()
        if amount <= self._cap:
            self._cap -= amount
            return True
        return False

    def reset(self):
        self._cap = self._burst
        self._now = self._timefunc()

    @property
    def capacity(self):
        self._replenish()
        return self._cap

File: receiver/tbf/tbf.go
package tbf

import (
	"time"
)

// TokenBucket rate limiter
type TokenBucket struct {
	// parameters of tbf
	rate  float32
	burst float32

	// state
	capacity float32
	lastts   time.Time
}

// New creates a new TokenBucket rate limiter
func New(rate, burst float32) *TokenBucket {
	return &TokenBucket{
		rate:     rate / float32(time.Second),
		burst:    burst,
		capacity: burst,
		lastts:   time.Now(),
	}
}

// Get gets amount tokens from the bucket, non-blocking.
// Returns true if tokens were available, and false otherwise
func (t *TokenBucket) Get(amount float32) bool {
	t.replenish()

	if t.capacity >= amount {
		t.capacity -= amount
		return true
	}

	return false
}

func (t *TokenBucket) replenish() {
	now := time.Now()
	t.capacity += t.rate * float32(now.Sub(t.lastts))
	if t.capacity > t.burst {
		t.capacity = t.burst
	}
	t.lastts = now
}
